module/apmfasthttp/context.go

```diff
@@ -0,0 +1,91 @@
+package apmfasthttp
+
+import (
+	"strings"
+
+	"github.com/savsgio/gotils/strconv"
+	"github.com/valyala/fasthttp"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/module/apmhttp"
+)
+
+func setRequestContext(tx *Transaction, ctx *fasthttp.RequestCtx) error {
+	req := &tx.req
+
+	if err := requestCtxToRequest(ctx, req, &tx.httpBody); err != nil {
+		return err
+	}
+
+	bc := tx.tracer.CaptureHTTPRequestBody(req)
+
+	tx.body = bc
+	tx.tx.Context.SetHTTPRequest(req)
+	tx.tx.Context.SetHTTPRequestBody(bc)
+
+	return nil
+}
+
+func setResponseContext(tx *Transaction) error { // nolint:unparam
+	statusCode := tx.httpCtx.Response.Header.StatusCode()
+
+	tx.tx.Result = apmhttp.StatusCodeResult(statusCode)
+	if !tx.tx.Sampled() {
+		return nil
+	}
+
+	headers := tx.req.Header
+	resetHTTPMap(headers)
+
+	tx.httpCtx.Response.Header.VisitAll(func(k, v []byte) {
+		sk := strconv.B2S(k)
+		sv := strconv.B2S(v)
+
+		headers.Set(sk, sv)
+	})
+
+	tx.tx.Context.SetHTTPResponseHeaders(headers)
+	tx.tx.Context.SetHTTPStatusCode(statusCode)
+
+	tx.body.Discard()
+
+	return nil
+}
+
+// StartTransaction returns a new Transaction with name,
+// created with tracer, and taking trace context from ctx.
+//
+// If the transaction is not ignored, the request and the request body
+// capturer will be returned with the transaction added to its context.
+func StartTransaction(tracer *apm.Tracer, name string, ctx *fasthttp.RequestCtx) (*Transaction, error) {
+	traceContext, ok := getRequestTraceparent(ctx, apmhttp.ElasticTraceparentHeader)
+	if !ok {
+		traceContext, ok = getRequestTraceparent(ctx, apmhttp.W3CTraceparentHeader)
+	}
+
+	if ok {
+		tracestateHeader := strconv.B2S(ctx.Request.Header.Peek(apmhttp.TracestateHeader))
+		traceContext.State, _ = apmhttp.ParseTracestateHeader(strings.Split(tracestateHeader, ",")...)
+	}
+
+	tx := acquireTransaction()
+	tx.tracer = tracer
+	tx.tx = tracer.StartTransactionOptions(name, "request", apm.TransactionOptions{TraceContext: traceContext})
+	tx.httpCtx = ctx
+
+	if err := setRequestContext(tx, ctx); err != nil {
+		tx.Close()
+
+		return nil, err
+	}
+
+	return tx, nil
+}
+
+// GetTransaction returns the transaction from the given ctx, if exist.
+func GetTransaction(ctx *fasthttp.RequestCtx) *Transaction {
+	if tx, ok := ctx.UserValue(TxKey).(*Transaction); ok {
+		return tx
+	}
+
+	return nil
+}
```

Comment:
It's important that `apm.TransactionFromContext` work with the RequestContext object, so `apm.StartSpan` (which takes a `context.Context`) just works. I think we can achieve this by overriding `internal/apmcontext.TransactionFromContext`: https://github.com/elastic/apm-agent-go/blob/65dbc25481c1f2a21125758c292cacd37c3dd98e/internal/apmcontext/context.go#L56

Something along these lines:

```go
func init() {
    origTransactionFromContext := apmcontext.TransactionFromContext
    apmcontext.TransactionFromContext = func(ctx context.Context) interface{} {
        tx, ok := ctx.Value(TxKey).(*apm.Transaction)
        if !ok {
            tx = origTransactionFromContext(ctx)
        }
        return tx
    }
}
```

Note that the function should return `*apm.Transaction`. I'd like to avoid having the Transaction wrapper type if possible -- see other comments.

Reply:
I'm afraid that fasthttp does not support `context.Context` for save data, only have the `UserValues` whose key must be a string.

So you could do:
```go
tx := apmfasthttp.GetTransaction(ctx) // returns the transaction wrapper (due to their reasons)
apmTx := tx.Tx() // returns the underliying APM transaction.
apmContext := apmcontext.TransactionFromContext(context.Background(), apmTx)
.....
```